### PR TITLE
Add multicast to udp socket

### PR DESCRIPTION
**Pull request by 28Smiles**
This should make it possible to implement mDNS for embassy-net, which in turn is essential for smart home

### CODE REVIEW

File: embassy-net/src/udp.rs [hunk 1]
```diff
@@ -26,13 +25,13 @@ pub enum Error {
     NoRoute,
 }
 
-pub struct UdpSocket<'a> {
-    stack: &'a RefCell<SocketStack>,
+pub struct UdpSocket<'a, D: Driver> {
```
Comment (Dirbaio) [inline embassy-net/src/udp.rs:28]: It is intentional that the socket is not generic on the driver. It prevents generics spam proliferation.

File: embassy-net/src/udp.rs [hunk 2]
```diff
@@ -143,8 +139,29 @@ impl<'a> UdpSocket<'a> {
     }
 }
 
-impl Drop for UdpSocket<'_> {
+#[cfg(feature = "igmp")]
+impl<'a, D: Driver + smoltcp::phy::Device + 'static> UdpSocket<'a, D> {
+    pub fn join_multicast_group<T>(&self, addr: T) -> Result<bool, smoltcp::iface::MulticastError>
```
Comment (Dirbaio) [inline embassy-net/src/udp.rs:144]: I don't think these methods should be in UdpSocket. Them being here implies the "joining" is scoped to the socket, ie if you drop the socket you un-join the group. This is not the case though.

these should be in `Stack` instead, mirroring smoltcp.

Comment (28Smiles) [reply embassy-net/src/udp.rs]: I made the previous helpers on Stack public and reverted the changes on the Socket.

**Review by Dirbaio — APPROVED** [review]
Thank you!

bors r+

**Pull request merged**